federation-js/src/joinSpec.ts

```diff
@@ -27,59 +27,83 @@ const JoinGraphDirective = new GraphQLDirective({
-    const noNumericFirstChar = alphaNumericUnderscoreOnly.match(/^[0-9]/)
+  function sanitizeGraphQLName(name: string) {
+    // replace all non-word characters (\W). Word chars are _a-zA-Z0-9
+    const alphaNumericUnderscoreOnly = name.replace(/[\W]/g, '_');
+    // prefix a digit in the first position with an _
+    const noNumericFirstChar = alphaNumericUnderscoreOnly.match(/^\d/)
       ? '_' + alphaNumericUnderscoreOnly
       : alphaNumericUnderscoreOnly;
-    const noUnderscoreNumericEnding = noNumericFirstChar.match(/_[0-9]+$/)
+    // suffix an underscore + digit in the last position with an _
+    const noUnderscoreNumericEnding = noNumericFirstChar.match(/_\d+$/)
       ? noNumericFirstChar + '_'
       : noNumericFirstChar;
 
     // toUpper not really necessary but follows convention of enum values
     const toUpper = noUnderscoreNumericEnding.toLocaleUpperCase();
+    return toUpper;
+  }
+
+  // duplicate enum values can occur due to sanitization and must be accounted for
+  // collect the duplicates in an array so we can uniquify them in a second pass.
+  const sanitizedNameToServiceDefinitions: Map<
+    string,
+    ServiceDefinition[]
+  > = new Map();
+  for (const service of sortedServiceList) {
+    const { name } = service;
+    const sanitized = sanitizeGraphQLName(name);
 
-    // Uniquifying post-transform
-    const nameCount = nameMap.get(toUpper);
-    if (nameCount) {
-      // Collision - bump counter by one
-      nameMap.set(toUpper, nameCount + 1);
-      const uniquified = `${toUpper}_${nameCount + 1}`;
-      // We also now need another entry for the name we just generated
-      nameMap.set(uniquified, 1);
-      graphNameToEnumValueName[name] = uniquified;
-      return uniquified;
+    const existingEntry = sanitizedNameToServiceDefinitions.get(sanitized);
+    if (existingEntry) {
+      sanitizedNameToServiceDefinitions.set(sanitized, [
+        ...existingEntry,
+        service,
+      ]);
     } else {
-      nameMap.set(toUpper, 1);
-      graphNameToEnumValueName[name] = toUpper;
-      return toUpper;
+      sanitizedNameToServiceDefinitions.set(sanitized, [service]);
     }
   }
 
+  // if no duplicates for a given name, add it as is
+  // if duplicates exist, append _{n} (index-1) to each duplicate in the array
+  const generatedNameToServiceDefinition: Record<
+    string,
+    ServiceDefinition
+  > = Object.create(null);
+  for (const [name, services] of sanitizedNameToServiceDefinitions) {
+    if (services.length === 1) {
+      generatedNameToServiceDefinition[name] = services[0];
+    } else {
+      for (const [index, service] of services.entries()) {
+        generatedNameToServiceDefinition[`${name}_${index + 1}`] = service;
+      }
+    }
+  }
+
+  const entries = Object.entries(generatedNameToServiceDefinition);
   return {
-    graphNameToEnumValueName,
+    graphNameToEnumValueName: Object.fromEntries(
+      entries.map(([name, service]) => [service.name, name]),
```

Comment:
I definitely think clarity would be improved by replacing `name` with `enumName` or `enumValueName` in both `map` calls.

Reply:
208a0415